cpp/include/kvikio/file_handle.hpp

```diff
@@ -549,44 +556,51 @@ class FileHandle {
-   * the actual size can be set prior to the stream I/O execution.
-   * @param file_offset Pointer to offset in the file from which to read. Unless otherwise set
+   * @param size_p Pointer to size in bytes to read. If the exact size is not known at the time of
+   * I/O submission, then you must set it to the maximum possible I/O size for that stream I/O.
+   * Later the actual size can be set prior to the stream I/O execution.
+   * @param file_offset_p Pointer to offset in the file from which to read. Unless otherwise set
    * using cuFileStreamRegister API, this value will not be evaluated until execution time.
-   * @param devPtr_offset Pointer to the offset relative to the bufPtr_base pointer from which to
+   * @param devPtr_offset_p Pointer to the offset relative to the bufPtr_base pointer from which to
    * write.
-   * @param bytes_written Pointer to the bytes read from file. This pointer should be a non-NULL
-   * value and *bytes_written set to 0. The bytes_written memory should be allocated with
+   * @param bytes_written_p Pointer to the bytes read from file. This pointer should be a non-NULL
+   * value and *bytes_written_p set to 0. The bytes_written_p memory should be allocated with
    * cuMemHostAlloc/malloc/mmap or registered with cuMemHostRegister.
-   * After successful execution of the operation in the stream, the value *bytes_written will
+   * After successful execution of the operation in the stream, the value *bytes_written_p will
    * contain either:
    *     - The number of bytes successfully read.
    *     - -1 on IO errors.
    *     - All other errors return a negative integer value of the CUfileOpError enum value.
    * @param stream CUDA stream in which to enqueue the operation. If NULL, make this operation
    * synchronous.
    */
-  inline void write_async(void* devPtr_base,
-                          std::size_t* size,
-                          off_t* file_offset,
-                          off_t* devPtr_offset,
-                          ssize_t* bytes_written,
-                          CUstream stream)
+  void write_async(void* devPtr_base,
+                   std::size_t* size_p,
+                   off_t* file_offset_p,
+                   off_t* devPtr_offset_p,
+                   ssize_t* bytes_written_p,
+                   CUstream stream)
   {
 #ifdef KVIKIO_CUFILE_STREAM_API_FOUND
-    CUFILE_TRY(cuFileAPI::instance().WriteAsync(
-      _handle, devPtr_base, size, file_offset, devPtr_offset, bytes_written, stream));
-    return;
-#else
-    throw CUfileException("cuFile's stream API isn't available, please build with CUDA v12.2+.");
+    if (kvikio::is_batch_and_stream_available() && !kvikio::defaults::compat_mode()) {
+      CUFILE_TRY(cuFileAPI::instance().WriteAsync(
+        _handle, devPtr_base, size_p, file_offset_p, devPtr_offset_p, bytes_written_p, stream));
+      return;
+    }
 #endif
+
```

Comment:
As with the `read` case, put this fallback in an `#else` block? Same rationale, so if that isn't the intended behaviour, feel free to ignore in the same way.